### ISSUE TITLE
Zookeeper remains in not ready state after zk pod delete

### ISSUE THREAD

**Issue by Raghuveera-Rao** [pravega/pravega-operator#158]
Deployed 
Pravega-Operator image: pravega/pravega-operator:0.3.2  
Zookeeper-operator image: pravega/zookeeper-operator:0.2.1
Ran small i/o  workload (Total - 4 readers, 3 writers, 120 events/sec, ~ 100 KB/sec IO) 

Environment details: PKS / K8 with medium cluster:

```
3 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
10 worker nodes @ xlarge.cpu(8 cpu, 8 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend 
[pravega-k8s-logging-2019-04-04-01-49-45.zip](https://github.com/pravega/pravega-operator/files/3042005/pravega-k8s-logging-2019-04-04-01-49-45.zip)

```

Deleted one pod of zookeeper and segment store
`
$ kubectl delete po/pravega-zk-1 po/pravega-pravega-segmentstore-0
`
pravega-pravega-segmentstore-0 came up in few minutes.
**But, pravega-zk-1 is still showing 0/1**

kubectl get po output is:

```
NAME                                                 READY     STATUS    RESTARTS   AGE
pravega-benchmark                                    1/1       Running   0          21h
pravega-bookie-0                                     1/1       Running   0          20h
pravega-bookie-1                                     1/1       Running   0          22h
pravega-bookie-2                                     1/1       Running   0          22h
pravega-bookie-3                                     1/1       Running   0          22h
pravega-operator-779879b48-8lhlx                     1/1       Running   0          22h
pravega-pravega-controller-c67d6b758-787c8           1/1       Running   1          22h
pravega-pravega-controller-c67d6b758-nfsll           1/1       Running   0          22h
pravega-pravega-segmentstore-0                       1/1       Running   0          19h
pravega-pravega-segmentstore-1                       1/1       Running   0          22h
pravega-pravega-segmentstore-2                       1/1       Running   0          22h
pravega-zk-0                                         1/1       Running   0          20h
pravega-zk-1                                         0/1       Running   0          27m
pravega-zk-2                                         1/1       Running   0          23h
shaka-zulu-nfs-client-provisioner-59d7f8f84c-wx8wl   1/1       Running   0          22h
zookeeper-operator-685bfcbbc5-qktpm                  1/1       Running   0          23h

```

kubectl describe po/pravega-zk-1:
```
`
Name:               pravega-zk-1
Namespace:          default
Priority:           0
PriorityClassName:  <none>
Node:               4e18637b-d138-44dc-89c0-096cbeb4af70/172.23.1.8
Start Time:         Thu, 04 Apr 2019 01:44:27 -0400
Labels:             app=pravega-zk
                    controller-revision-hash=pravega-zk-65b97d6f5b
                    kind=ZookeeperMember
                    statefulset.kubernetes.io/pod-name=pravega-zk-1
Annotations:        <none>
Status:             Running
IP:                 172.24.2.4
Controlled By:      StatefulSet/pravega-zk
Containers:
  zookeeper:
    Container ID:  docker://b30e6b0978d6d0bd1c7122c588c54750f778c416338a7342c0451dcc03117715
    Image:         emccorp/zookeeper:3.5.4-beta-operator
    Image ID:      docker-pullable://emccorp/zookeeper@sha256:c4656ca1e0103b1660a978fa8b2bfecfae5baf6746bb1839360821abd209082f
    Ports:         2181/TCP, 2888/TCP, 3888/TCP
    Host Ports:    0/TCP, 0/TCP, 0/TCP
    Command:
      /usr/local/bin/zookeeperStart.sh
    State:          Running
      Started:      Thu, 04 Apr 2019 01:44:47 -0400
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  2Gi
    Requests:
      cpu:        500m
      memory:     1Gi
    Liveness:     exec [zookeeperLive.sh] delay=10s timeout=10s period=10s #success=1 #failure=3
    Readiness:    exec [zookeeperReady.sh] delay=10s timeout=10s period=10s #success=1 #failure=3
    Environment:  <none>
    Mounts:
      /conf from conf (rw)
      /data from data (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-mprtn (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  data-pravega-zk-1
    ReadOnly:   false
  conf:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      pravega-zk-configmap
    Optional:  false
  default-token-mprtn:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-mprtn
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                  Age   From                                           Message
  ----     ------                  ----  ----                                           -------
  Normal   Scheduled               30m   default-scheduler                              Successfully assigned default/pravega-zk-1 to 4e18637b-d138-44dc-89c0-096cbeb4af70
  Warning  FailedAttachVolume      30m   attachdetach-controller                        Multi-Attach error for volume "pvc-61440e79-55df-11e9-b5dc-00505694490a" Volume is already exclusively attached to one node and can't be attached to another
  Normal   SuccessfulAttachVolume  29m   attachdetach-controller                        AttachVolume.Attach succeeded for volume "pvc-61440e79-55df-11e9-b5dc-00505694490a"
  Normal   Pulling                 29m   kubelet, 4e18637b-d138-44dc-89c0-096cbeb4af70  pulling image "emccorp/zookeeper:3.5.4-beta-operator"
  Normal   Pulled                  29m   kubelet, 4e18637b-d138-44dc-89c0-096cbeb4af70  Successfully pulled image "emccorp/zookeeper:3.5.4-beta-operator"
  Normal   Created                 29m   kubelet, 4e18637b-d138-44dc-89c0-096cbeb4af70  Created container
  Normal   Started                 29m   kubelet, 4e18637b-d138-44dc-89c0-096cbeb4af70  Started container
  Warning  Unhealthy               29m   kubelet, 4e18637b-d138-44dc-89c0-096cbeb4af70  Readiness probe failed: Zookeeper service is ready to be upgraded from observer to participant.
+ source /conf/env.sh
++ DOMAIN=pravega-zk-headless.default.svc.cluster.local
++ QUORUM_PORT=2888
++ LEADER_PORT=3888
++ CLIENT_HOST=pravega-zk-client
++ CLIENT_PORT=2181
+ source /usr/local/bin/zookeeperFunctions.sh
++ set -ex
++ hostname -s
+ HOST=pravega-zk-1
+ DATA_DIR=/data
+ MYID_FILE=/data/myid
+ LOG4J_CONF=/conf/log4j-quiet.properties
++ ++ nc 127.0.0.1 echo 2181
ruok
+ OK=imok
+ [[ imok == \i\m\o\k ]]
+ [[ pravega-zk-1 =~ (.*)-([0-9]+)$ ]]
+ NAME=pravega-zk
+ ORD=1
+ [[ 1 -eq 0 ]]
+ set +e
++ zkConnectionString
++ set +e
++ nslookup pravega-zk-client
++ [[ 0 -eq 1 ]]
++ set -e
++ echo pravega-zk-client:2181
+ ZKURL=pravega-zk-client:2181
+ set -e
++ cat /data/myid
+ MYID=2
++ java -Dlog4j.configuration=file:/conf/log4j-quiet.properties -jar /root/zu.jar get-role pravega-zk-client:2181 2
Connecting to Zookeeper pravega-zk-client:2181
+ ROLE=observer
+ [[ observer == \p\a\r\t\i\c\i\p\a\n\t ]]
+ [[ observer == \o\b\s\e\r\v\e\r ]]
+ echo 'Zookeeper service is ready to be upgraded from observer to participant.'
+ ROLE=participant
++ zkConfig
++ echo 'pravega-zk-1.pravega-zk-headless.default.svc.cluster.local:2888:3888:participant;2181'
+ ZKCONFIG='pravega-zk-1.pravega-zk-headless.default.svc.cluster.local:2888:3888:participant;2181'
+ java -Dlog4j.configuration=file:/conf/log4j-quiet.properties -jar /root/zu.jar remove pravega-zk-client:2181 2
Connecting to Zookeeper pravega-zk-client:2181
+ sleep 1
+ java -Dlog4j.configuration=file:/conf/log4j-quiet.properties -jar /root/zu.jar add pravega-zk-client:2181 2 'pravega-zk-1.pravega-zk-headless.default.svc.cluster.local:2888:3888:participant;2181'
Connecting to Zookeeper pravega-zk-client:2181
Error performing zookeeper reconfiguration:
org.apache.zookeeper.KeeperException$NewConfigNoQuorum: KeeperErrorCode = NewConfigNoQuorum
           at org.apache.zookeeper.KeeperException.create(KeeperException.java:110)
           at org.apache.zookeeper.KeeperException.create(KeeperException.java:54)
           at org.apache.zookeeper.ZooKeeper.internalReconfig(ZooKeeper.java:2993)
           at org.apache.zookeeper.admin.ZooKeeperAdmin.reconfigure(ZooKeeperAdmin.java:174)
           at io.pravega.zookeeper.MainKt.reconfigure(Main.kt:99)
           at io.pravega.zookeeper.MainKt.runAdd(Main.kt:122)
           at io.pravega.zookeeper.MainKt.main(Main.kt:29)
  Warning  Unhealthy  28m  kubelet, 4e18637b-d138-44dc-89c0-096cbeb4af70  Readiness probe failed: + source /conf/env.sh
++ DOMAIN=pravega-zk-headless.default.svc.cluster.local
++ QUORUM_PORT=2888
++ LEADER_PORT=3888
++ CLIENT_HOST=pravega-zk-client
++ CLIENT_PORT=2181
+ source /usr/local/bin/zookeeperFunctions.sh
++ set -ex
++ hostname -s
+ HOST=pravega-zk-1
+ DATA_DIR=/data
+ MYID_FILE=/data/myid
+ LOG4J_CONF=/conf/log4j-quiet.properties
++ nc 127.0.0.1 2181
++ echo ruok
+ OK=imok
+ [[ imok == \i\m\o\k ]]
+ [[ pravega-zk-1 =~ (.*)-([0-9]+)$ ]]
+ NAME=pravega-zk
+ ORD=1
+ [[ 1 -eq 0 ]]
+ set +e
++ zkConnectionString
++ set +e
++ nslookup pravega-zk-client
++ [[ 0 -eq 1 ]]
++ set -e
++ echo pravega-zk-client:2181
+ ZKURL=pravega-zk-client:2181
+ set -e
++ cat /data/myid
+ MYID=2
++ java -Dlog4j.configuration=file:/conf/log4j-quiet.properties -jar /root/zu.jar get-role pravega-zk-client:2181 2
Connecting to Zookeeper pravega-zk-client:2181
Server not found in zookeeper config
+ ROLE=
  Warning  Unhealthy  4m (x145 over 29m)  kubelet, 4e18637b-d138-44dc-89c0-096cbeb4af70  Readiness probe failed: + source /conf/env.sh
++ DOMAIN=pravega-zk-headless.default.svc.cluster.local
++ QUORUM_PORT=2888
++ LEADER_PORT=3888
++ CLIENT_HOST=pravega-zk-client
++ CLIENT_PORT=2181
+ source /usr/local/bin/zookeeperFunctions.sh
++ set -ex
++ hostname -s
+ HOST=pravega-zk-1
+ DATA_DIR=/data
+ MYID_FILE=/data/myid
+ LOG4J_CONF=/conf/log4j-quiet.properties
++ echo ruok
++ nc 127.0.0.1 2181
+ OK=imok
+ [[ imok == \i\m\o\k ]]
+ [[ pravega-zk-1 =~ (.*)-([0-9]+)$ ]]
+ NAME=pravega-zk
+ ORD=1
+ [[ 1 -eq 0 ]]
+ set +e
++ zkConnectionString
++ set +e
++ nslookup pravega-zk-client
++ [[ 0 -eq 1 ]]
++ set -e
++ echo pravega-zk-client:2181
+ ZKURL=pravega-zk-client:2181
+ set -e
++ cat /data/myid
+ MYID=2
++ java -Dlog4j.configuration=file:/conf/log4j-quiet.properties -jar /root/zu.jar get-role pravega-zk-client:2181 2
Connecting to Zookeeper pravega-zk-client:2181
Server not found in zookeeper config
+ ROLE=


```

**Comment by Raghuveera-Rao** [comment]
[pravega-k8s-logging-2019-04-04-01-49-45.zip](https://github.com/pravega/pravega-operator/files/3042006/pravega-k8s-logging-2019-04-04-01-49-45.zip)

attached logs

**Comment by Raghuveera-Rao** [comment]
describe pod log mentioned above has "Warning  Unhealthy  28m  kubelet, 4e18637b-d138-44dc-89c0-096cbeb4af70  Readiness probe failed"
What does unhealthy state mean ?

**Comment by pbelgundi** [comment]
This seems to be related to the zookeeper configuration problem and should be resolved with fix for this issue: https://github.com/pravega/zookeeper-operator/issues/66

**Comment by pbelgundi** [comment]
This is fixed with zk-operator version 0.2.3
